### PR TITLE
feat: [DHIS2-12364] Add RuleActionExtractDataMatrixValue function

### PR DESCRIPTION
**Pull request by Balcan**
This is a draft.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -291,7 +291,7 @@
         <dependency>
             <groupId>org.hisp.dhis.parser</groupId>
             <artifactId>dhis-antlr-expression-parser</artifactId>
-            <version>1.0.22</version>
+            <version>1.0.23</version>
```
Comment (enricocolasante) [inline pom.xml]: Antlr parser version is already to 1.0.24

```suggestion
            <version>1.0.24</version>
```

File: src/main/java/org/hisp/dhis/rules/gs1/GS1BaseValueFormatter.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package org.hisp.dhis.rules.gs1;
+
+public abstract class GS1BaseValueFormatter
+        implements GS1ValueFormatter
+{
+
+    public static GS1ValueFormatter getFormatterFromValue( String value )
+    {
+        if( value == null )
+            throw new IllegalArgumentException( "Can't extract data from null value" );
```
Comment (enricocolasante) [inline src/main/java/org/hisp/dhis/rules/gs1/GS1BaseValueFormatter.java:10]: What happens if value.length < 3?
Should we handle this case?

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        return extractDataMatrixValue( value, gs1Key );
+    }
+
+    @Override
+    public Object getDescription( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+
+        visitor.castStringVisit( ctx.expr( 0 ) );
+
```
Comment (enricocolasante) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java:61]: The getDescription method is also used to validate. It means that if a description is returned the system is confident that the expression is valid.
I think we should call the extractDataMatrixValue method here as well

**Review by enricocolasante — COMMENTED** [review]
Nice job!
I have just some concerns on the validation side.
Let me know what you think about that

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java (hunk 3 above)
Comment (Balcan) [reply src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java:61]: replacing this line with:
`        extractDataMatrixValue( visitor.castStringVisit( ctx.expr( 1 ) ), visitor.castStringVisit( ctx.expr( 0 ) ) );
`
Would be enough?

Comment (enricocolasante) [reply src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java:61]: Yes, I think so.
With this change, if the function is not valid, the description will show it

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**